File: TypeScript/Algorithm/Algorithm II/15. 3Sum.ts
function threeSum(nums: number[]): number[][] {
  const NUMS_LENGTH: number = nums.length;

  const uniqueSolution: number[][] = [];

  nums.sort((a, b) => a - b);

  for (let i: number = 0; i < NUMS_LENGTH; i++) {
    if (i > 0 && nums[i] === nums[i - 1]) 
      continue;

    let left: number = i + 1;
    let right: number = NUMS_LENGTH - 1;

    while (left < right) {
      const currentSum: number = nums[i] + nums[left] + nums[right];

      if (currentSum > 0) 
        right--;
      
      else if (currentSum < 0) 
        left++;
      
      else {
        uniqueSolution.push([nums[i], nums[left], nums[right]]);
        left++;

        while (left < right && nums[left] === nums[left - 1]) 
          left++;
      }
    }
  }

  return uniqueSolution;
}

File: Java/Algorithm/Algorithm II/15. 3Sum.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {

  /**
   * Given an array of integers, finds all unique triplets that sum to zero (0).
   *
   * @param nums An array of integers where triplets are to be found.
   * @return A list of lists containing unique triplets that sum to zero (0).
   */
  public List<List<Integer>> threeSum(int[] nums) {
    int NUMS_LENGTH = nums.length;

    List<List<Integer>> uniqueSolution = new ArrayList<>();

    Arrays.sort(nums);

    for (int i = 0; i < NUMS_LENGTH - 2; i++) {
      if (i > 0 && nums[i] == nums[i - 1]) 
        continue;

      int left = i + 1;
      int right = NUMS_LENGTH - 1;

      while (left < right) {
        int currentSum = nums[i] + nums[left] + nums[right];

        if (currentSum < 0) 
          left++; 
          
        else if (currentSum > 0) 
          right--; 
        
        else {
          List<Integer> solution = new ArrayList<>();
          solution.add(nums[i]);
          solution.add(nums[left]);
          solution.add(nums[right]);
          uniqueSolution.add(solution);

          left++;

          while (left < right && nums[left] == nums[left - 1]) 
            left++;
        }
      }
    }

    return uniqueSolution;
  }
}
